File: 0301-0400/341-350/349E_Intersection_of_Two_Arrays/java/main.java
// 349. Intersection of Two Arrays
// https://leetcode.com/problems/intersection-of-two-arrays/
// T.C.: O(n + m)
// S.C.: O(n)

// java main.java

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

class Main {

  public static void main(String[] args) {
    Solution solution = new Solution();

    int[] nums1 = { 1, 2, 2, 1 };
    int[] nums2 = { 2, 2 };
    int[] expected = { 2 };
    int[] result = solution.intersection(nums1, nums2);
    System.out.println(result + " " + (Arrays.equals(expected, result) ? "true" : "false"));

    nums1 = new int[] { 4, 9, 5 };
    nums2 = new int[] { 9, 4, 9, 8, 4 };
    expected = new int[] { 4, 9 };
    result = solution.intersection(nums1, nums2);
    System.out.println(result + " " + (Arrays.equals(expected, result) ? "true" : "false"));
  }
}

class Solution {

  public int[] intersection(int[] nums1, int[] nums2) {
    if (nums1.length < nums2.length) {
      return this.intersection(nums2, nums1);
    }
    Set<Integer> set1 = new HashSet<>();
    Set<Integer> set2 = new HashSet<>();
    for (int num : nums1) {
      set1.add(num);
    }
    for (int num : nums2) {
      if (set1.contains(num)) {
        set2.add(num);
      }
    }
    int[] result = new int[set2.size()];
    int index = 0;
    for (int num : set2) {
      result[index++] = num;
    }
    return result;
  }
}

File: 0301-0400/341-350/349E_Intersection_of_Two_Arrays/rust/src/main.rs
// 349. Intersection of Two Arrays
// https://leetcode.com/problems/intersection-of-two-arrays/
// T.C.: O(n + m)
// S.C.: O(n)
use colored::Colorize;
use colored::ColoredString;
struct Solution;

use std::collections::HashSet;

impl Solution {
    pub fn intersection(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
        if nums1.len() < nums2.len() {
            return Self::intersection(nums2, nums1);
        }
        let mut result = HashSet::new();
        let mut nums = HashSet::new();
        for num in nums1.clone() {
            nums.insert(num);
        }
        for num in nums2.clone() {
            if nums.contains(&num) {
                result.insert(num);
            }
        }
        return Vec::from_iter(result);
    }
}

fn main() {
    let mut nums1 = vec![1, 2, 2, 1];
    let mut nums2 = vec![2, 2];
    let mut expected = vec![2];
    test(nums1, nums2, expected);

    nums1 = vec![4, 9, 5];
    nums2 = vec![9, 4, 9, 8, 4];
    expected = vec![4, 9];
    test(nums1, nums2, expected);
}

fn test(nums1: Vec<i32>, nums2: Vec<i32>, expected: Vec<i32>) {
    let result = Solution::intersection(nums1.clone(), nums2.clone());

    println!("\nresult: {}\nassert: {}", array_colored(result.clone()), bool_colored(result == expected));
}

fn bool_colored(value: bool) -> ColoredString {
    let text;
    if value {
        text = value.to_string().green();
    } else {
        text = value.to_string().red();
    }
    return text;
}

fn array_colored(nums: Vec<i32>) -> ColoredString {
    let text = format!(
        "[{}]",
        nums
            .iter()
            .map(|x| x.to_string())
            .collect::<Vec<String>>()
            .join(", ")
    );
    return text.yellow();
}
